File: backends/spring-java/src/main/java/dev/stackverse/backend/audit/AdminAuditController.java
package dev.stackverse.backend.audit;

import static dev.stackverse.backend.common.RequestValidation.requireValidPaging;

import com.fasterxml.jackson.annotation.JsonInclude;
import dev.stackverse.backend.common.PageResponse;
import jakarta.persistence.criteria.Predicate;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Map;
import java.util.UUID;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import tools.jackson.databind.ObjectMapper;

@JsonInclude(JsonInclude.Include.NON_NULL)
record AuditEntryResponse(
    UUID id,
    String actor,
    String action,
    String targetType,
    String targetId,
    Map<String, Object> detail,
    Instant createdAt
) {
}

@RestController
@RequestMapping("/api/v1/admin/audit-log")
@PreAuthorize("hasRole('admin')")
public class AdminAuditController {
    private final AuditRepository repository;
    private final ObjectMapper objectMapper;

    public AdminAuditController(AuditRepository repository, ObjectMapper objectMapper) {
        this.repository = repository;
        this.objectMapper = objectMapper;
    }

    @GetMapping
    public PageResponse<AuditEntryResponse> list(
        @RequestParam(name = "actor", required = false) String actor,
        @RequestParam(name = "action", required = false) String action,
        @RequestParam(name = "targetType", required = false) String targetType,
        @RequestParam(name = "targetId", required = false) String targetId,
        @RequestParam(name = "from", required = false) Instant from,
        @RequestParam(name = "to", required = false) Instant to,
        @RequestParam(name = "page", defaultValue = "0") int page,
        @RequestParam(name = "size", defaultValue = "20") int size
    ) {
        requireValidPaging(page, size);
        Specification<AuditEntry> specification = (root, query, cb) -> {
            var predicates = new ArrayList<Predicate>();
            if (actor != null) {
                predicates.add(cb.equal(root.get("actor"), actor));
            }
            if (action != null) {
                predicates.add(cb.equal(root.get("action"), action));
            }
            if (targetType != null) {
                predicates.add(cb.equal(root.get("targetType"), targetType));
            }
            if (targetId != null) {
                predicates.add(cb.equal(root.get("targetId"), targetId));
            }
            if (from != null) {
                predicates.add(cb.greaterThanOrEqualTo(root.get("createdAt"), from));
            }
            if (to != null) {
                predicates.add(cb.lessThanOrEqualTo(root.get("createdAt"), to));
            }
            return cb.and(predicates.toArray(Predicate[]::new));
        };
        return PageResponse.of(
            repository.findAll(specification, PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"))),
            this::toResponse
        );
    }

    private AuditEntryResponse toResponse(AuditEntry entry) {
        return new AuditEntryResponse(
            entry.getId(),
            entry.getActor(),
            entry.getAction(),
            entry.getTargetType(),
            entry.getTargetId(),
            readDetail(entry.getDetail()),
            entry.getCreatedAt()
        );
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> readDetail(String detail) {
        if (detail == null) {
            return null;
        }
        return objectMapper.readValue(detail, Map.class);
    }
}

File: backends/spring-kotlin/src/main/kotlin/dev/stackverse/backend/audit/AdminAuditController.kt
package dev.stackverse.backend.audit

import com.fasterxml.jackson.annotation.JsonInclude
import dev.stackverse.backend.common.PageResponse
import dev.stackverse.backend.common.requireValidPaging
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.data.jpa.domain.Specification
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import tools.jackson.databind.ObjectMapper
import java.time.Instant
import java.util.UUID

@JsonInclude(JsonInclude.Include.NON_NULL)
data class AuditEntryResponse(
    val id: UUID,
    val actor: String,
    val action: String,
    val targetType: String,
    val targetId: String,
    val detail: Map<String, Any?>?,
    val createdAt: Instant,
)

@RestController
@RequestMapping("/api/v1/admin/audit-log")
@PreAuthorize("hasRole('admin')")
class AdminAuditController(
    private val repository: AuditRepository,
    private val objectMapper: ObjectMapper,
) {

    @GetMapping
    fun list(
        @RequestParam actor: String?,
        @RequestParam action: String?,
        @RequestParam targetType: String?,
        @RequestParam targetId: String?,
        @RequestParam from: Instant?,
        @RequestParam to: Instant?,
        @RequestParam(defaultValue = "0") page: Int,
        @RequestParam(defaultValue = "20") size: Int,
    ): PageResponse<AuditEntryResponse> {
        requireValidPaging(page, size)
        val specification = Specification.allOf(
            buildList<Specification<AuditEntry>> {
                actor?.let { add(Specification { root, _, cb -> cb.equal(root.get<String>("actor"), it) }) }
                action?.let { add(Specification { root, _, cb -> cb.equal(root.get<String>("action"), it) }) }
                targetType?.let { add(Specification { root, _, cb -> cb.equal(root.get<String>("targetType"), it) }) }
                targetId?.let { add(Specification { root, _, cb -> cb.equal(root.get<String>("targetId"), it) }) }
                from?.let { add(Specification { root, _, cb -> cb.greaterThanOrEqualTo(root.get("createdAt"), it) }) }
                to?.let { add(Specification { root, _, cb -> cb.lessThanOrEqualTo(root.get("createdAt"), it) }) }
            },
        )
        val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"))
        return PageResponse.of(repository.findAll(specification, pageable)) { it.toResponse() }
    }

    private fun AuditEntry.toResponse() = AuditEntryResponse(
        id = id,
        actor = actor,
        action = action,
        targetType = targetType,
        targetId = targetId,
        detail = detail?.let { objectMapper.readValue(it, Map::class.java).mapKeys { (k, _) -> k.toString() } },
        createdAt = createdAt,
    )
}
